File: 06 - DAT203.1x - Data Science Essentials/Labs/Mod5/PlotOutliers.R
vis.outlier <- function(col1 = 'citympg'){
  require(ggplot2) # make sure ggplot2 is loaded
  ## convert character columns to factors for plotting
  auto.price[, "outlier"] <- as.factor(auto.price[, "outlier"])    
  auto.price[, "fueltype"] <- as.factor(auto.price[, "fueltype"])    
  title = paste('Plot of', col1, 'vs.lnprice') # character string title
  ggplot(auto.price, aes_string(col1, 'lnprice')) +
    geom_point(aes(color = outlier, 
                   shape = fueltype, 
                   alpha = 0.5, size = 4)) +
    ggtitle(title)
}

id.outlier <- function(df){
  ## Use ifelse to filter for outliers
  df[, "outlier"] <- ifelse(df[,"enginesize"] > 190 | 
                              df[, "weight"] > 3500| 
                              df[, "citympg"] > 40, '1', '0')   
  df
}

## ID and plot outliers
auto.price <- maml.mapInputPort(1) # read autos data frame from port 1
require(dplyr) # make sure dplyr is loaded
## set the columns to plot
plot.cols <- c("weight",
               "enginesize",
               "citympg")
auto.price <- id.outlier(auto.price) # id outliers
lapply(plot.cols, vis.outlier)  # plot the results
out <- auto.price %>% filter(outlier == 1) # return the outliers
maml.mapOutputPort("out") ## Output the data frame

File: 06 - DAT203.1x - Data Science Essentials/Labs/Mod5/PlotOutliers.py
def id_outlier(df):
    ## Create a vector of 0 of length equal to the number of rows
    temp = [0] * df.shape[0]
    ## test each outlier condition and mark with a 1 as required
    for i, x in enumerate(df['enginesize']):
        if (x > 190): temp[i] = 1 
    for i, x in enumerate(df['weight']):
        if (x > 3500): temp[i] = 1 
    for i, x in enumerate(df['citympg']):
        if (x > 40): temp[i] = 1      
    df['outlier'] = temp # append a column to the data frame
    return df

def auto_scatter_outlier(df, plot_cols):
    import matplotlib.pyplot as plt
    outlier = [0,0,1,1] # Vector of outlier indicators
    fuel = ['gas','diesel','gas','diesel']  # vector of fuel types
    color = ['DarkBlue','DarkBlue','Red','Red'] # vector of color choices for plot
    marker = ['x','o','o','x'] # vector of shape choices for plot
    for col in plot_cols: # loop over the columns
        fig = plt.figure(figsize=(6, 6))
        ax = fig.gca()
        ## Loop over the zip of the four vectors an subset the data and
        ## create the plot using the aesthetics provided
        for o, f, c, m in zip(outlier, fuel, color, marker):
            temp = df.ix[(df['outlier'] == o) & (df['fueltype'] == f)]           
            if temp.shape[0] > 0:                    
                temp.plot(kind = 'scatter', x = col, y = 'lnprice' , 
                           ax = ax, color = c, marker = m)                                 
        ax.set_title('Scatter plot of lnprice vs. ' + col)
        fig.savefig('scatter_' + col + '.png')
    return plot_cols      

def azureml_main(df): 
    import matplotlib
    matplotlib.use('agg')
    ## Define the plot columns
    plot_cols = ["weight",
               "enginesize",
               "citympg"]
    df = id_outlier(df)  # mark outliers       
    auto_scatter_outlier(df, plot_cols) # create plots
    df = df[df.outlier == 1] # filter for outliers
    return df
